internal/configfsm/fsm.go

```diff
@@ -0,0 +1,309 @@
+	InProgressToCompleted Trigger = "InProgress->Completed"
+
+	// Transitions to OutOfDate state
+	InProgressToOutOfDate      Trigger = "InProgress->OutOfDate"
+	TimedOutToOutOfDate        Trigger = "TimedOut->OutOfDate"
+	ClusterNotReadyToOutOfDate Trigger = "ClusterNotReady->OutOfDate"
+	CompletedToOutOfDate       Trigger = "Completed->OutOfDate"
+
+	// Self transitions
+	MissingToMissing                 Trigger = "Missing->Missing"
+	ClusterNotReadyToClusterNotReady Trigger = "ClusterNotReady->ClusterNotReady"
+	InProgressToInProgress           Trigger = "InProgress->InProgress"
+	TimedOutToTimedOut               Trigger = "TimedOut->TimedOut"
+	CompletedToCompleted             Trigger = "Completed->Completed"
+	OutOfDateToOutOfDate             Trigger = "OutOfDate->OutOfDate"
+
+	// States
+	Missing         = "Missing"
+	Start           = "Start"
+	ClusterNotReady = "ClusterNotReady"
+	InProgress      = "InProgress"
+	TimedOut        = "TimedOut"
+	Completed       = "Completed"
+	OutOfDate       = "OutOfDate"
+)
+
+// RunFSM Runs the state machine as much as it can by triggering all Self transitions
+func RunFSM(ctx context.Context, fsm *stateless.StateMachine, fsmHelper FsmHelper) (aState any, err error) {
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	switch aState {
+	case Start:
+		err = fsm.Fire(StartToMissing, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Missing:
+		err = fsm.Fire(MissingToMissing, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case ClusterNotReady:
+		err = fsm.Fire(ClusterNotReadyToClusterNotReady, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case InProgress:
+		err = fsm.Fire(InProgressToInProgress, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case TimedOut:
+		err = fsm.Fire(TimedOutToTimedOut, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Completed:
+		err = fsm.Fire(CompletedToCompleted, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case OutOfDate:
+		err = fsm.Fire(OutOfDateToOutOfDate, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	}
+	// Update the current state
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	return aState, nil
+}
+
+// InitFSM Initialize the state machine algorithm
+func InitFSM(state string) (fsm *stateless.StateMachine, err error) {
+	fsm = stateless.NewStateMachine(state)
+	fsm.Configure(Start).Permit(StartToMissing, Missing)
```

Comment:
Does this mean that `missing` is checked first? If there are no policies and Cluster is not ready, will the status condition set to Missing instead of ClusterNotReady?

Reply:
Thanks for explanation!